### PR TITLE
kernel: Add ticks to low priority threads that arbitrate zero threads

### PR DESCRIPTION
**Pull request by GPUCode**
Supersedes https://github.com/citra-emu/citra/pull/6977

Turns out that unconditionally adding many ticks to svc::ArbitrateAddress caused audio desync in some games and broke others like the Japanese Rail Sim games that are very timitng sensitive. Looking at the kernel diassembly, it appears that Nintendo has placed a small workaround for the case we are dealing with, mainly when low priority threads attempt to signal and end up waking nobody. By implementing this workaround instead we can achieve similar speed gains to before without the afformentioned side-effects. The number of cycles has been taken from the HOS kernel.

There is a small difference however: the real kernel checks for priorities bigger or equal than 50, while here we check for 49. The affected FMV thread has a priority of 49 in citra which barely misses this number. I presume this is likely a scheduling bug as our scheduler isn't the best. Future kernel refactors will shed more light into this, but for now reducing this by one does not seem to break anything from my testing.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7096)

### CODE REVIEW

File: src/core/hle/kernel/address_arbiter.h [hunk 1]
```diff
@@ -65,11 +65,11 @@ class AddressArbiter final : public Object, public WakeupCallback {
     void WaitThread(std::shared_ptr<Thread> thread, VAddr wait_address);
 
     /// Resume all threads found to be waiting on the address under this address arbiter
-    void ResumeAllThreads(VAddr address);
+    u64 ResumeAllThreads(VAddr address);
```
Comment (innovation2013) [inline src/core/hle/kernel/address_arbiter.h:68]: The declaration is inconsistent with the defined return value type

**Comment by innovation2013** [comment]
This PR brings a huge improvement in fmv speed. It is obvious that the fmv speed of Monster Hunter 3G has been increased from the original 10fps to a stable 30fps.

File: src/core/hle/kernel/address_arbiter.cpp [hunk 2]
```diff
@@ -107,17 +109,24 @@ ResultCode AddressArbiter::ArbitrateAddress(std::shared_ptr<Thread> thread, Arbi
     switch (type) {
 
     // Signal thread(s) waiting for arbitrate address...
-    case ArbitrationType::Signal:
+    case ArbitrationType::Signal: {
         // Negative value means resume all threads
```
Comment (SachinVin) [inline src/core/hle/kernel/address_arbiter.cpp:113]: comment should be moved down

File: src/core/hle/kernel/address_arbiter.cpp [hunk 3]
```diff
@@ -107,17 +109,24 @@ ResultCode AddressArbiter::ArbitrateAddress(std::shared_ptr<Thread> thread, Arbi
...
             // Resume first N threads
-            for (int i = 0; i < value; i++)
-                ResumeHighestPriorityThread(address);
+            for (s32 i = 0; i < value; i++) {
+                num_threads += ResumeHighestPriorityThread(address);
+            }
+        }
+        // Prevents lag from low priority threads that spam svcArbitrateAddress and wake no threads
+        // The value below is taken from official HOS kernel
+        if (num_threads == 0 && thread->current_priority >= 49) {
```
Comment (SachinVin) [inline src/core/hle/kernel/address_arbiter.cpp:125]: are both `49` and `1614` from the kernel?

File: src/core/hle/kernel/svc.cpp [hunk 4]
```diff
@@ -851,8 +851,10 @@ ResultCode SVC::WaitSynchronizationN(s32* out, VAddr handles_address, s32 handle
 
         // If a timeout value of 0 was provided, just return the Timeout error code instead of
         // suspending the thread.
-        if (nano_seconds == 0)
+        if (nano_seconds == 0) {
+            system.GetRunningCore().GetTimer().AddTicks(2000);
```
Comment (SachinVin) [inline src/core/hle/kernel/svc.cpp:855]: Where does 2000 come from?

File: src/core/hle/kernel/address_arbiter.cpp (hunk 2 above)
Comment (GPUCode) [reply src/core/hle/kernel/address_arbiter.cpp]: This was included by mistake, should be in a separate PR

File: src/core/hle/kernel/svc.cpp (hunk 4 above)
Comment (GPUCode) [reply src/core/hle/kernel/svc.cpp]: This was included by mistake, should be in a separate PR

File: src/core/hle/kernel/address_arbiter.cpp (hunk 3 above)
Comment (SachinVin) [reply src/core/hle/kernel/address_arbiter.cpp:127]: I think it makes sense to add the reason in the comments

**Review by Steveice10 — APPROVED** [review]
Still needs some more canary testing but the code LGTM

**Review by FearlessTobi — APPROVED** [review]
Code lgtm

**Pull request merged**